File: not/teams.c
// This program reads in a JSON file containing information about Genshin Impact characters and their stats
// It then calculates the best team compositions based on those stats, and outputs the top 5 teams

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// define struct for each character
struct character {
    char key[20]; // name of character
    int level; // level of character
    int constellation; // number of constellations unlocked
    int ascension; // current ascension level
    struct talent {
        int auto; // level of auto attack talent
        int skill; // level of skill talent
        int burst; // level of burst talent
    } talent;
};

// define function to read in JSON file
void read_json(char* filename, struct character* characters) {
    // read in file
    FILE* fp = fopen(filename, "r");
    if (fp == NULL) {
        printf("Error: could not open file\n");
        exit(1);
    }
    char buffer[100];
    fgets(buffer, 100, fp);
    fclose(fp);
    
    // parse JSON and store in character structs
    // ...
}

// define function to calculate team scores
int calculate_score(struct character* team) {
    // calculate total score for the team based on character stats
    // ...
}

// define function to output top 5 teams
void output_top_teams(struct character** teams) {
    // output the top 5 teams in order of score
    // ...
}

int main() {
    // read in JSON file and store characters in array of structs
    struct character characters[4];
    read_json("characters.json", characters);
    
    // generate all possible team compositions
    struct character* teams[5];
    // ...
    
    // calculate scores for each team
    int scores[5];
    for (int i = 0; i < 5; i++) {
        scores[i] = calculate_score(teams[i]);
    }
    
    // output top 5 teams based on score
    output_top_teams(teams);
    
    return 0;
}

File: teams.py
import json


with open('characters.json') as f:
    data = json.load(f)

characters = data['characters']


teams = []


def calculate_team_score(team):
    score = 0
    for character in team:
        score += character['talent']['auto'] + character['talent']['skill'] + character['talent']['burst']
    return score
# TODO ugrade the algorithm

for i in range(len(characters)):
    for j in range(i+1, len(characters)):
        for k in range(j+1, len(characters)):
            for l in range(k+1, len(characters)):
                team = [characters[i], characters[j], characters[k], characters[l]]

                if 'RaidenShogun' in [c['key'] for c in team] and 'Eula' in [c['key'] for c in team]:
                    teams.append(team)


teams.sort(key=calculate_team_score, reverse=True)


for i, team in enumerate(teams[:5]):
    print(f'Team {i+1}:')
    for character in team:
        print(f'{character["key"]}, level {character["level"]}, constellation {character["constellation"]}, ascension {character["ascension"]}, talent: auto {character["talent"]["auto"]}, skill {character["talent"]["skill"]}, burst {character["talent"]["burst"]}')
    print(f'Total score: {calculate_team_score(team)}\n')